File: src/com/urbanairship/octobot/Metrics.scala
package com.urbanairship.octobot

import java.util.ArrayList
import java.util.HashMap
import java.util.LinkedList

object Metrics {

    // Keep track of all tasks we've seen executed.
    val instrumentedTasks = new ArrayList[String]()
    
    // Keep track of average task throughput (last 10k runs per task).
    val executionTimes = new HashMap[String, LinkedList[Long]]()

    // Keep track of total successes by task.
    val taskSuccesses = new HashMap[String, Int]()

    // Keep track of total failures by task.
    val taskFailures = new HashMap[String, Int]()
    
    // Keep track of total retries by task.
    val taskRetries = new HashMap[String, Int]()

    val metricsLock = new Object()

    
    // Updates internal metrics following task execution.
    def update(task: String, time: Long, status: Boolean, retries: Int) {
        metricsLock.synchronized {
            if (!instrumentedTasks.contains(task)) instrumentedTasks.add(task)

            updateExecutionTimes(task, time)
            updateTaskRetries(task, retries)
            updateTaskResults(task, status)
        }
    }


    // Update the list of execution times, keeping the last 10,000 per task.
    def updateExecutionTimes(task: String, time: Long) {
        if (!executionTimes.containsKey(task)) {
            val timeList = new LinkedList[Long]()
            timeList.addFirst(time)
            executionTimes.put(task, timeList)
        } else {
            val timeList = executionTimes.get(task)
            if (timeList.size() == 10000) timeList.removeLast()
            timeList.addFirst(time)
            executionTimes.put(task, timeList)
        }
    }


    // Update the number of times this task has been retried.
    def updateTaskRetries(task: String, retries: Int) {
        if (retries > 0) {
            if (!taskRetries.containsKey(task)) {
                taskRetries.put(task, retries)
            } else {
                var retriesForTask = taskRetries.get(task)
                retriesForTask += retries
                taskRetries.put(task, retriesForTask)
            }
        }   
    }


    // Update the number of times this task has succeeded or failed.
    def updateTaskResults(task: String, status: Boolean) {
        if (status == true) {
            if (!taskSuccesses.containsKey(task)) {
                taskSuccesses.put(task, 1)
            } else {
                var success = taskSuccesses.get(task)
                taskSuccesses.put(task, success + 1)
            }
        } else {
            if (!taskFailures.containsKey(task)) {
                taskFailures.put(task, 1)
            } else {
                var failure = taskFailures.get(task)
                taskFailures.put(task, failure + 1)
            }
        }
    }

}

File: src/main/java/com/urbanairship/octobot/Metrics.java
package com.urbanairship.octobot;

import com.yammer.metrics.core.Counter;
import com.yammer.metrics.core.MetricName;
import com.yammer.metrics.core.MetricsRegistry;
import com.yammer.metrics.core.Timer;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.concurrent.TimeUnit;

public class Metrics {

    protected static final MetricsRegistry registry = new MetricsRegistry();

    // Updates internal metrics following task execution.
    public static void update(String task, long time, boolean status, int retries) {
        updateExecutionTimes(task, time);
        updateTaskRetries(task, retries);
        updateTaskResults(task, status);
    }


    // Update the list of execution times, keeping the last 10,000 per task.
    private static void updateExecutionTimes(String task, long time) {
        MetricName timerName = new MetricName("Octobot", "Metrics", task + "Timer");
        Timer timer = registry.newTimer(timerName, TimeUnit.MILLISECONDS, TimeUnit.SECONDS);
        timer.update(time, TimeUnit.MILLISECONDS);
    }


    // Update the number of times this task has been retried.
    private static void updateTaskRetries(String task, int retries) {
        MetricName counterRetriesName = new MetricName("Octobot", "Metrics", task + "Retries");
        Counter counterRetries = registry.newCounter(counterRetriesName);
        counterRetries.inc();
    }


    // Update the number of times this task has succeeded or failed.
    private static void updateTaskResults(String task, boolean status) {
        if (status == true) {
            MetricName counterSuccessName = new MetricName("Octobot", "Metrics", task + "Success");
            Counter counterSuccess = registry.newCounter(counterSuccessName);
            counterSuccess.inc();
        } else {
            MetricName counterFailureName = new MetricName("Octobot", "Metrics", task + "Failure");
            Counter counterFailure = registry.newCounter(counterFailureName);
            counterFailure.inc();
        }
    }
    
}

